pkg/tasks/ccm_csi_migration.go

```diff
@@ -189,6 +233,107 @@ func updateKubeletConfigInternal(s *state.State, node *kubeoneapi.HostConfig, co
 	return nil
 }
 
+func ccmMigrationUpdateKubeletConfigFile(s *state.State) error {
+	// Grab the Kubelet configuration file from the node
+	sshfs := s.Runner.NewFS()
+	f, err := sshfs.Open(kubeletConfigFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	buf, err := io.ReadAll(f)
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal and update the config
+	kubeletConfig, err := unmarshalKubeletConfig(buf)
+	if err != nil {
+		return err
+	}
+
+	if kubeletConfig.FeatureGates == nil {
+		kubeletConfig.FeatureGates = map[string]bool{}
+	}
+	if s.ShouldEnableCSIMigration() {
+		featureGates, _, fgErr := s.Cluster.CSIMigrationFeatureGates(s.ShouldUnregisterInTreeCloudProvider())
+		if fgErr != nil {
+			return fgErr
+		}
+		for k, v := range featureGates {
+			kubeletConfig.FeatureGates[k] = v
+		}
+	}
+
+	// Update the config on the node
+	buf, err = marshalKubeletConfig(kubeletConfig)
+	if err != nil {
+		return err
+	}
+
+	fw, ok := f.(sshiofs.ExtendedFile)
+	if !ok {
+		return errors.New("file is not writable")
+	}
+
+	if err = fw.Truncate(0); err != nil {
+		return err
+	}
+
+	if _, err = fw.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+
+	if _, err = io.Copy(fw, bytes.NewBuffer(buf)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func ccmMigrationUpdateKubeletFlags(s *state.State) error {
+	sshfs := s.Runner.NewFS()
+	f, err := sshfs.Open(kubeadmEnvFlagsFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	buf, err := io.ReadAll(f)
+	if err != nil {
+		return err
+	}
+
+	kubeletFlags, err := unmarshalKubeletFlags(buf)
+	if err != nil {
+		return err
+	}
+
+	kubeletFlags["--cloud-provider"] = "external"
+	delete(kubeletFlags, "--cloud-config")
+
+	buf = marshalKubeletFlags(kubeletFlags)
+	fw, ok := f.(sshiofs.ExtendedFile)
+	if !ok {
+		return errors.New("file is not writable")
+	}
+
+	if err = fw.Truncate(0); err != nil {
+		return err
+	}
+
+	if _, err = fw.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+
+	if _, err = io.Copy(fw, bytes.NewBuffer(buf)); err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
Not sure if it's worth it, but `ccmMigrationUpdateKubeletFlags()` and `ccmMigrationUpdateKubeletConfigFile()` are practically identical and only have very slight differences in behaviour. Maybe it make sense to extrapolate into a `updateRemoteFile(state *State, file string, modifier func(content string) string) error` (pseudocode)?

Reply:
I really like this idea. Addressed in a follow-up commit.